R/assign.R

```diff
@@ -0,0 +1,388 @@
+#'       "IA (Intra-arterial)",
+#'       "",
+#'       "Non-standard",
+#'       "random_value",
+#'       "IJ (Intra-articular)",
+#'       "TRANSDERMAL",
+#'       "OPHTHALMIC"
+#'     )
+#'   )
+#'
+#' assign_ct(
+#'   raw_dataset = md1,
+#'   raw_variable = "MDRTE",
+#'   study_ct = study_ct,
+#'   target_sdtm_variable = "CMROUTE",
+#'   target_sdtm_variable_codelist_code = "C66729"
+#'   )
+#'
+#' cm_inter <-
+#'   tibble::tibble(
+#'     oak_id = 1:14,
+#'     raw_source = "MD1",
+#'     PATIENT_NUM = 101:114,
+#'     CMTRT = c(
+#'       "BABY ASPIRIN",
+#'       "CORTISPORIN",
+#'       "ASPIRIN",
+#'       "DIPHENHYDRAMINE HCL",
+#'       "PARCETEMOL",
+#'       "VOMIKIND",
+#'       "ZENFLOX OZ",
+#'       "AMITRYPTYLINE",
+#'       "BENADRYL",
+#'       "DIPHENHYDRAMINE HYDROCHLORIDE",
+#'       "TETRACYCLINE",
+#'       "BENADRYL",
+#'       "SOMINEX",
+#'       "ZQUILL"
+#'     ),
+#'     CMINDC = c(
+#'       NA,
+#'       "NAUSEA",
+#'       "ANEMIA",
+#'       "NAUSEA",
+#'       "PYREXIA",
+#'       "VOMITINGS",
+#'       "DIARHHEA",
+#'       "COLD",
+#'       "FEVER",
+#'       "LEG PAIN",
+#'       "FEVER",
+#'       "COLD",
+#'       "COLD",
+#'       "PAIN"
+#'     ),
+#'     CMROUTE = c(
+#'       "ORAL",
+#'       "ORAL",
+#'       NA,
+#'       "ORAL",
+#'       "ORAL",
+#'       "ORAL",
+#'       "INTRAMUSCULAR",
+#'       "INTRA-ARTERIAL",
+#'       NA,
+#'       "NON-STANDARD",
+#'       "RANDOM_VALUE",
+#'       "INTRA-ARTICULAR",
+#'       "TRANSDERMAL",
+#'       "OPHTHALMIC"
+#'     )
+#'   )
+#'
+#' assign_ct(
+#'   raw_dataset = md1,
+#'   raw_variable = "MDRTE",
+#'   study_ct = study_ct,
+#'   target_sdtm_variable = "CMROUTE",
+#'   target_sdtm_variable_codelist_code = "C66729",
+#'   target_dataset = cm_inter,
+#'   merge_to_topic_by = c("oak_id","raw_source","PATIENT_NUM")
+#'   )
+#'
+#' @name assign
+NULL
+
+#' @export
+#' @rdname assign
+assign_no_ct <- function(raw_dataset,
+                         raw_variable,
+                         target_sdtm_variable,
+                         target_dataset = raw_dataset,
+                         merge_to_topic_by = NULL) {
+  sdtm_assign(
+    raw_dat = raw_dataset,
+    raw_var = raw_variable,
+    tgt_var = target_sdtm_variable,
+    tgt_dat = target_dataset,
+    by = merge_to_topic_by
+  )
+}
+
+#' Derive an SDTM variable with controlled terminology
+#'
+#' [assign_ct()] maps a variable in a source dataset to a target SDTM variable
+#' following controlled terminology recoding.
+#'
+#' @param raw_dataset The raw dataset.
+#' @param raw_variable The raw variable.
+#' @param target_sdtm_variable The target SDTM variable.
+#' @param target_dataset Target dataset. By default the same as `raw_dataset`.
+#' @param merge_to_topic_by If `target_dataset` is different than `raw_dataset`,
+#'   then this parameter defines keys to use in the join between `raw_dataset`
+#'   and `target_dataset`.
+#' @param study_ct Study controlled terminology specification.
```

Comment:
I think your question hints at a more deep issue here. I think we should have a proper S3 (or even the new S7) classed object that represents controlled terminology mappings. It could be self-validating in the sense of ensuring that there are always the needed variables involved in the recoding. Then, functions like `assign_ct()` would only need to assert that `ct` is indeed an object of a certain type, such that all these other assumptions about having this and that column could be taken for granted.